pkg/pillar/cmd/domainmgr/domainmgr.go

```diff
@@ -1116,6 +1149,154 @@ func lookupDomainConfig(ctx *domainContext, key string) *types.DomainConfig {
+	for vmUUID := range ctx.resourcesUsedByVMs {
+		vmToCheck := ctx.resourcesUsedByVMs[vmUUID]
+		// Need the Status here only to get the proper Key to find the corresponding Config
+		vmStatus := lookupDomainStatusByUUID(ctx, vmUUID)
+		if vmStatus == nil {
+			//Most probably the status has not been created yet
+			continue
+		}
+		if vmStatus.VmConfig.CPUsPinned {
+			// We need to check only non-pinned VMs
+			continue
+		}
+		vmConfig := lookupDomainConfig(ctx, vmStatus.Key())
+		if vmConfig == nil {
+			// Strange situation. Should be investigated if happens.
+			log.Warnf("Failed to find the DomainConfig corresponding to a DomainStatus (%s)", vmStatus.DisplayName)
+			continue
+		}
+		cpuSetChanged := false
+		for cpuToCheck, used := range vmToCheck.CPUSet {
+			if cpuToCheck == cpu && used {
+				// Found previously assigned CPU. Have to exclude it
+				vmToCheck.CPUSet[cpu] = false
+				cpuSetChanged = true
+			}
+		}
+		// If at least one change in the CPU set of the VM happened, adapt the corresponding cgroup
+		if cpuSetChanged {
+			vmConfig.VmConfig.CPUs = constructNonPinnedCpumaskString(ctx)
+			err := setCgroupCpuset(vmConfig.GetTaskName(), vmConfig.VmConfig.CPUs)
+			if err != nil {
+				return fmt.Errorf("failed to redistribute CPUs between VMs, can affect the inter-VM isolation")
+			}
+			log.Functionf("The CPU set for %s has been changed to: %s", vmConfig.DisplayName, vmConfig.CPUs)
+		}
+	}
+	return nil
+}
+
+func assignAndPinCPU(ctx *domainContext, config *types.DomainConfig, cpu int) {
+	log.Functionf("Assign pinned CPU %d to %s", cpu, config.DisplayName)
+	ctx.resourcesUsedByVMs[config.UUIDandVersion.UUID].CPUSet[cpu] = true
+	ctx.hostCpusPinned[cpu] = true
+	addToMask(cpu, &config.VmConfig.CPUs)
+}
+
+func assignCPU(ctx *domainContext, config *types.DomainConfig, cpu int) {
+	log.Functionf("Assign non-pinned CPU %d to %s", cpu, config.DisplayName)
+	ctx.resourcesUsedByVMs[config.UUIDandVersion.UUID].CPUSet[cpu] = true
+	addToMask(cpu, &config.VmConfig.CPUs)
+}
+
+func assignCPUs(ctx *domainContext, config *types.DomainConfig) error {
+	var cpusAssignedNum = 0
+	if config.VmConfig.CPUsPinned { // Pin the CPU
+		for i := ctx.cpusReservedForEve; i < ctx.hostCpusNum; i++ {
+			pinned, _ := ctx.hostCpusPinned[i]
+			// Is it pinned at the moment?
+			if !pinned { // No, it's not used at the moment
+				assignAndPinCPU(ctx, config, i)
+				cpusAssignedNum++
+			}
+			if cpusAssignedNum == config.VmConfig.VCpus {
+				break
+			}
+		}
+		// Check,that all the necessary CPUs were found
+		if cpusAssignedNum != config.VmConfig.VCpus {
+			log.Warnf("Failed to allocate necessary amount of CPUs (assigned %d, need %d)", cpusAssignedNum, config.VmConfig.VCpus)
+			// Clean up already assign CPUs
+			for cpuToCheck, pinned := range ctx.resourcesUsedByVMs[config.UUIDandVersion.UUID].CPUSet {
+				if pinned {
+					ctx.hostCpusPinned[cpuToCheck] = false
```

Comment:
Should we be concered that the shared set (hostCpusPinned) can become empty?
I wonder whether we should always hold back one CPU here.
Alternatively, we would need to switch the shared set to use the EVE CPU?

Reply:
We always have the first `cpusReservedForEve` CPUs in a non-pinned state.

Reply:
But do those get used/shared by the guests which are not requesting pinning? 
If they do its fine; just need to add some comments explaining that the shared set is never empty because of this.

Reply:
The logic is reimplemented in the scope of the CPUAllocator